elastic/elastic-agent
# Collect process (beats) stdout and stderr

## What does this PR do?

The Elastic Agent now collects the `strout` and `stderr `of the (sub)process it runs.

## Why is it important?

When application crashes (panics), the "crash/panic" information is lost as they log directly to files and by default the stdout and stderr of a sub process is connected to `os.DevNull`.

## Checklist




## How to test this PR locally

Panic a beat and check the logs are there :)

There is the [`agent-88-collest-beats-stdout-stderr`](https://github.com/elastic/beats/tree/agent-88-collest-beats-stdout-stderr) beats branch has a modified filebeat that exposes a [HTTP server](https://github.com/elastic/beats/blob/agent-88-collest-beats-stdout-stderr/x-pack/filebeat/httppanic/httppanic.go#L18) which will [panic](https://github.com/elastic/beats/blob/46b5d90be40a31519854b4b68e09be0b002c5596/libbeat/monitoring/report/log/log.go#L168) and [log to stdout and stderr](https://github.com/elastic/beats/blob/agent-88-collest-beats-stdout-stderr/x-pack/filebeat/httppanic/httppanic.go#L23) upon request.

Just compile the agent and the beats together, run the agent with any policy that will install and run a filebeat (filebeat_monitoring also works).
Look for logs like `starting HTTP panic server on port :424X`, where X is a random number chosen at startup. The agent should already log it, but filebeat will [log it](https://github.com/elastic/beats/blob/agent-88-collest-beats-stdout-stderr/x-pack/filebeat/httppanic/httppanic.go#L36) as well.
Once you have it, you can:

```shell
curl localhost:424X # will make filebeat log to stdout and stderr
curl -X PANIC localhost:424X # will make filebeat panic
```

## Related issues

- Closes #88 

## Logs

```text
{"log.level":"info","@timestamp":"2022-05-13T16:19:28.760Z","log.origin":{"file.name":"process/stdlogger.go","file.line":50},"message":"filebeat stdout: \"starting HTTP panic server on port :4248\\n\"","agent.console.name":"filebeat","agent.console.type":"stdout","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-05-13T16:19:47.210Z","log.origin":{"file.name":"process/stdlogger.go","file.line":50},"message":"filebeat stdout: \"logging to stdout before PANIC panic!\\n\"","agent.console.name":"filebeat","agent.console.type":"stdout","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2022-05-13T16:20:01.802Z","log.origin":{"file.name":"process/stdlogger.go","file.line":50},"message":"filebeat stderr: \"panic: HTTP Panic server said to panic!\\n\\ngoroutine 65 [running]:\\ngithub.com/elastic/beats/v7/libbeat/monitoring/report/log.(*reporter).snapshotLoop(0xc000724640)\\n\\t/go/src/github.com/elastic/beats/libbeat/monitoring/report/log/log.go:169 +0x67d\\ngithub.com/elastic/beats/v7/libbeat/monitoring/report/log.MakeReporter.func1()\\n\\t/go/src/github.com/elastic/beats/libbeat/monitoring/report/log/log.go:134 +0x5a\\ncreated by github.com/elastic/beats/v7/libbeat/monitoring/report/log.MakeReporter\\n\\t/go/src/github.com/elastic/beats/libbeat/monitoring/report/log/log.go:132 +0x23e\\n\"","agent.console.name":"filebeat","agent.console.type":"stderr","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-05-20T10:31:02.518Z","log.origin":{"file.name":"process/stdlogger.go","file.line":50},"message":"filebeat stdout: \"starting HTTP panic server on port :4247\\n\"","agent.console.name":"filebeat","agent.console.type":"stdout","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-05-20T10:31:02.663Z","log.origin":{"file.name":"process/stdlogger.go","file.line":50},"message":"filebeat_monitoring stdout: \"starting HTTP panic server on port :4249\\n\"","agent.console.name":"filebeat_monitoring","agent.console.type":"stdout","ecs.version":"1.6.0"}
```

![Screenshot from 2022-05-20 12-22-41](https://user-images.githubusercontent.com/203669/169508419-88537488-f53e-491a-9d0e-16a3a6303c77.png)
